### PR TITLE
Added a JSON payload formatter

### PR DESCRIPTION
**Pull request by SgtPepperLHCB**
See https://github.com/Nordstrom/kafka-connect-lambda/issues/20

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -102,7 +137,7 @@ To make sure our Lambda works, invoke it directly and view the result payload in
 aws lambda invoke --function-name example-function --payload '{"value": "my example"}' result.txt
 ```
 
-The function simply sends the `payload` back to you in `result.txt`.
+The function simply sends the `payload` back to you in `result.txt` as serialized json.
```
Comment (dylanmei) [inline README.md]: It's not "serialized JSON" so much as whatever string you passed in. So, this change is not necessary.

Comment (dylanmei) [reply README.md]: I see, you've edited the tutorial. This kind of work seems off-topic.

File: docker-compose.yml [hunk 2]
```diff
@@ -46,33 +46,34 @@ services:
...
-
-    - AWS_PROFILE
-    - AWS_ACCESS_KEY_ID
-    - AWS_SECRET_ACCESS_KEY
-    volumes:
-    - ~/.aws:/root/.aws
-    - ./target/plugin:/opt/connectors
-    - ./config/log4j.properties:/etc/log4j.properties
-    depends_on: [broker]
+  # NB: run connect locally in stand-alone mode to debug
```
Comment (dylanmei) [inline docker-compose.yml]: Please don't change this file. Let's keep it simple.

File: src/main/java/com/nordstrom/kafka/connect/formatters/JsonPayloadFormatter.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+import static java.util.Collections.emptyMap;
+
+public class JsonPayloadFormatter implements PayloadFormatter, Configurable {
+  class SchemaVisiblityStrategy {
+    private boolean all = false;
+    private boolean none = false;
+    String visibility;
+
+    void configure(Map<String, ?> configs, String key) {
+      final Object visibility = configs.get(key);
```
Comment (dylanmei) [inline src/main/java/com/nordstrom/kafka/connect/formatters/JsonPayloadFormatter.java]: An enumeration / clearly-named single-value to represent the distrinct values is going to be more succinct

File: src/main/java/com/nordstrom/kafka/connect/formatters/JsonPayloadFormatter.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+      }
+
+      return mapper.writeValueAsString(p);
+    } catch (JsonProcessingException e) {
+      throw new PayloadFormattingException(e);
+    }
+  }
+
+  public String formatBatch(final Collection<SinkRecord> records) {
+    return "[{\"hello\": \"world\"}]";
```
Comment (dylanmei) [inline src/main/java/com/nordstrom/kafka/connect/formatters/JsonPayloadFormatter.java]: This seems incomplete. ;)

File: README.md (hunk 1 above)
Comment (SgtPepperLHCB) [reply README.md]: Just tried to make reformatting the payload in the lambda logs a little easier by making it serialized json rather than python dictionary serialization.

Comment (dylanmei) [reply README.md]: The intent of the guide and the example configs was to keep it very simple. If I could do it over again, I would use "hello world" instead of `{"value": "my example"}` as an example payload.

File: docker-compose.yml (hunk 2 above)
Comment (SgtPepperLHCB) [reply docker-compose.yml]: Restored connect node

File: config/worker.properties [hunk 5]
```diff
@@ -3,9 +3,13 @@ bootstrap.servers=localhost:9092
 plugin.path=target/plugin
 offset.storage.file.filename=/tmp/connect.offsets
 
-key.converter=org.apache.kafka.connect.storage.StringConverter
+# default to Avro/Avro for key/value converters
```
Comment (dylanmei) [inline config/worker.properties]: This is also not necessary -- not here to teach people to configure Kafka Connect. If we want to improve the quality of the README and guide, let's do that in another PR.

File: src/test/java/com/nordstrom/kafka/connect/lambda/LambdaSinkConnectorConfigTest.java [hunk 6]
```diff
@@ -64,9 +64,116 @@ public void sampleConfig() {
...
+
+    @Test
+    public void jsonPayloadFormatterNoSchemasConfig() {
+        LambdaSinkConnectorConfig config = new LambdaSinkConnectorConfig(
+            new HashMap<String, String>() {
+                {
+                    put("name", "test-connector");
+                    put("aws.lambda.function.arn", "test-function");
+                    put("payload.formatter.class", JsonPayloadFormatter.class.getCanonicalName());
+                    put("payload.formatter.schemas.enable", "false");
```
Comment (dylanmei) [inline src/test/java/com/nordstrom/kafka/connect/lambda/LambdaSinkConnectorConfigTest.java]: Let's update this and any similar checks to `payload.formatter.(key|value).schema.visiblity`.

File: README.md [hunk 7]
```diff
@@ -33,6 +33,20 @@ In addition to the standard [Kafka Connect connector configuration](https://kafk
 | `retries.max` | No | `5` | Maximum number of invocation retries |
 | `topics` | Yes | | Comma-delimited Kafka topics names to sink |
 
+## Formatters
```
Comment (dylanmei) [inline README.md]: Consider omitting second and third introduction sentence, as this is repeated later, when we've introduced concept of schema visibility.

File: README.md [hunk 8]
```diff
@@ -33,6 +33,20 @@ In addition to the standard [Kafka Connect connector configuration](https://kafk
...
+
+| Property | Required | Default value | Description |
+|:---------|:---------|:--------|:------------|
+| `payload.formatter.class` | No | `com.nordstrom.kafka.connect.formatters.PlainPayloadFormatter` | Specifies the formatter to use. |
+| `payload.formatter.key.schema.visibility` | No | `min` | Determines whether schema (if present) is included. Only applies to JsonPayloadFormatter |
+| `payload.formatter.value.schema.visibility` | No | `min` | Determines whether schema (if present) is included. Only applies to JsonPayloadFormatter |
+
+Schema `visibility` can have values of `none`, `min`, and `all` (default=`min`) and only applies to the `JsonPayloadFormatter`.  Including schema in the payload can result in very large messages.  `min` will include the schema `name` only and `version`.  `none` will null out these values.  `all` will include the entire serialized schema.
```
Comment (dylanmei) [inline README.md]: Consider re-arranging this to denote the included formatters, and then discuss `schema.visiblity` concept: 

> The connector includes two `payload.formatter.class` implementations:
> - `com.nordstrom.kafka.connect.formatters.PlainPayloadFormatter`
> - `com.nordstrom.kafka.connect.formatters.JsonPayloadFormatter`

> Including the full schema information in the invocation payload may result in very large messages. Therefore, use the `schema.visibility` key and value properties to control how much of the schema, if present, is included in the invocation payload: `none`, `all`, or `none`, `min`, and `all` (default=`min`). These settings apply to the `JsonPayloadFormatter`; the `PlainPayloadFormatter` always includes `min` schema information.

Alternately we can just add this behavior into the `PlainPayloadFormatter` and reduce this documentation complexity altogether.

File: README.md [hunk 9]
```diff
@@ -65,18 +79,39 @@ By supplying `com.nordstrom.kafka.connect.auth.AWSAssumeRoleCredentialsProvider`
...
+    "topic": "example-stream_string-avro",
     "partition": 1,
     "offset": 0,
-    "key": "",
+    "timestamp": 1567723257583,
+    "timestampTypeName": "CreateTime"
+}
+```
+
+### JsonPayloadFormatter
```
Comment (dylanmei) [inline README.md]: For this and the previous example, what are the _critical_ schema-related connector properties to get us to this output? Consider including this info for both sections, or each one individually:

```
key.converter=org.apache.kafka.connect.storage.StringConverter
value.converter=io.confluent.connect.avro.AvroConverter
value.converter.schema.registry.url=http://localhost:8080
payload.formatter.class=com.nordstrom.kafka.connect.formatters.JsonPayloadFormatter
```

File: src/main/java/com/nordstrom/kafka/connect/formatters/Payload.java [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+  private String keySchemaVersion;
+  private VType value;
+  private String valueSchemaName;
+  private String valueSchemaVersion;
+  private String topic;
+  private int partition;
+  private long offset;
+  private long timestamp;
+  private String timestampTypeName;
+  //TODO add header Map<String,String>.
```
Comment (dylanmei) [inline src/main/java/com/nordstrom/kafka/connect/formatters/Payload.java]: Let's omit this information from this PR. I think we have a good grasp on it.

File: src/main/java/com/nordstrom/kafka/connect/formatters/Payload.java [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+  // key/value since Header has a key, value and schema.
+
+  protected Payload() {
+  }
+
+  public Payload(final SinkRecord record) {
+//    this.key = record.key() == null ? "" : record.key().toString();
+    if (record.keySchema() != null) {
+      this.keySchemaName = record.keySchema().name();
+      //NB: may be null if not set in schema.
```
Comment (dylanmei) [inline src/main/java/com/nordstrom/kafka/connect/formatters/Payload.java]: This code speaks for itself; Let's omit this kind of comment.

File: src/main/java/com/nordstrom/kafka/connect/lambda/LambdaSinkConnectorConfig.java [hunk 12]
```diff
@@ -336,7 +374,7 @@ public static ConfigDef configDef() {
             this.documentation = documentation;
         }
 
-        String getValue() {
+            String getValue() {
```
Comment (dylanmei) [inline src/main/java/com/nordstrom/kafka/connect/lambda/LambdaSinkConnectorConfig.java]: Stray space?

File: src/main/java/com/nordstrom/kafka/connect/formatters/JsonPayloadFormatter.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    public boolean isNone() {
+      return none;
+    }
+  }
+
+  private final ObjectMapper mapper = new ObjectMapper();
+  private final JsonConverter converter = new JsonConverter();
+  private final JsonConverter converterSansSchema = new JsonConverter();
```
Comment (dylanmei) [inline src/main/java/com/nordstrom/kafka/connect/formatters/JsonPayloadFormatter.java]: I'd like to see  just one JsonCoverter class member; configured during `configure`. This allows us to avoid the `includeSchema` check during `deserailize`.

Comment (SgtPepperLHCB) [reply src/main/java/com/nordstrom/kafka/connect/formatters/JsonPayloadFormatter.java]: The formatter/converter is configured during task `start()` and then re-used for `put()`.  We need to be prepared to handle both at runtime and/or also need separate converters for key and value since their visibility is independent.

Comment (dylanmei) [reply src/main/java/com/nordstrom/kafka/connect/formatters/JsonPayloadFormatter.java]: I see, thanks for explaining.

File: README.md (hunk 1 above)
Comment (dylanmei) [reply README.md]: On second glance, this is still very simple and understandable, which is what we wanted.

File: README.md [hunk 14]
```diff
@@ -81,6 +81,39 @@ The default invocation payload is a JSON representation of a [SinkRecord](https:
 
 Example payloads for a SinkRecord with a simple string for a key and an Avro record for the value are illustrated below.
 
+The Avro schema for this is:
```
Comment (dylanmei) [inline README.md]: Consider replacing this with header and description
> ### Avro schema
> This simple schema record describes our "hello world" message.
